components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -566,6 +569,36 @@ private static boolean isTenantActive(String tenantDomain) throws IdentityOAuth2
         }
     }
 
+    /**
+     * Resolve the user from ancestor organizations and set the user ID if the authenticated user id is not populated.
+     *
+     * @param authenticatedUser              The authenticated user object.
+     * @param authenticatedSubjectIdentifier The authenticated subject identifier.
+     */
+    public static void setUserIdIfNotExist(AuthenticatedUser authenticatedUser, String authenticatedSubjectIdentifier) {
+
+        if (StringUtils.equals(authenticatedUser.getLoggableUserId(), authenticatedUser.toFullQualifiedUsername())) {
+            if (authenticatedUser.getAuthenticatedSubjectIdentifier() == null) {
+                return;
+            }
+            String userId = authenticatedSubjectIdentifier.split("@")[0];
```

Comment:
Subject identifier would be different if the subject identifier is explicitly set for an app. Hence this workaround will not help.